### PR TITLE
[FEATURE] Implement embedding removal methods for Weaviate embedding

### LINKED ISSUE #1171: [FEATURE] Implement embedding removal methods for Weaviate embedding store
`WeaviateEmbeddingStore` should have `remove(String id)`, `removeAll(Collection<String> ids)`, `removeAll(Filter filter)` and `removeAll()` methods implemented, similarly to `PgVectorEmbeddingStore`.

### PR DESCRIPTION
**Pull request by kiratkumar47**
## Issue
fix #1171 

## Change
Implemented EmbeddingStore removal method (remove(String id), removeAll(Collection<String> ids)) and removeAll()) in WeaviateEmbeddingStore.

## General checklist
- [X] There are no breaking changes
- [ ] I have added unit and integration tests for my change
- [ ] I have manually run all the unit and integration tests in the module I have added/changed, and they are all green
- [ ] I have manually run all the unit and integration tests in the [core](https://github.com/langchain4j/langchain4j/tree/main/langchain4j-core) and [main](https://github.com/langchain4j/langchain4j/tree/main/langchain4j) modules, and they are all green

- [ ] I have added/updated the [documentation](https://github.com/langchain4j/langchain4j/tree/main/docs/docs)
- [ ] I have added an example in the [examples repo](https://github.com/langchain4j/langchain4j-examples) (only for "big" features)

### CODE REVIEW

**Comment by kiratkumar47** [comment]
I have implemented the remove(String id), removeAll(Collection<String> ids), and removeAll() methods, but I am not sure how to implement the removeAll(Filter filter) method. I would greatly appreciate it if you could direct me in the right direction so that I can implement this method too.

**Comment by langchain4j** [comment]
@kiratkumar47 thank you! Please add/copy tests (e.g from `PgVectorEmbeddingStoreRemoveIT`) for the existing methods.

For the filter, I guess here is a good place to start: https://weaviate.io/developers/weaviate/api/graphql/filters

**Comment by langchain4j** [comment]
You can also check how `PgVectorEmbeddingStore` implements it. `MilvusMetadataFilterMapper` and `ElasticsearchMetadataFilterMapper` might also help

**Comment by kiratkumar47** [comment]
Hello,

I have implemented the removeAll(Filter filter) method. I copied the tests to the WeaviateEmbeddingStoreRemoveIT class from PgVectorEmbeddingStoreRemoveIT, but remove_all_by_filter() is failing. Upon debugging, I found that the public String add(Embedding embedding, TextSegment textSegment) method in WeaviateEmbeddingStore doesn't store the provided metadata in Weaviate. I don't know if this is intentional. It would be great if you could review my code and provide feedback.

File: langchain4j-weaviate/src/test/java/dev/langchain4j/store/embedding/weaviate/WeaviateEmbeddingStoreRemoveIT.java [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static dev.langchain4j.internal.Utils.randomUUID;
+import static dev.langchain4j.store.embedding.filter.MetadataFilterBuilder.metadataKey;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+@Testcontainers
+public class WeaviateEmbeddingStoreRemoveIT {
```
Comment (langchain4j) [inline langchain4j-weaviate/src/test/java/dev/langchain4j/store/embedding/weaviate/WeaviateEmbeddingStoreRemoveIT.java]:
```suggestion
class LocalWeaviateEmbeddingStoreRemoveIT {
```

File: langchain4j-weaviate/src/main/java/dev/langchain4j/store/embedding/weaviate/WeaviateEmbeddingStore.java [hunk 2]
```diff
@@ -146,6 +149,49 @@ public List<String> addAll(List<Embedding> embeddings, List<TextSegment> embedde
...
+    @Override
+    public void removeAll(Filter filter) {
+        ensureNotNull(filter, "filter");
+        WhereFilter whereFilter = WeaviateMetadataFilterMapper.map(filter);
+        client.batch().objectsBatchDeleter()
+                .withClassName(objectClass)
+                .withWhere(WhereFilter.builder()
+                        .path("metadata.id")
+                        .operator(Operator.ContainsAny)
+                        .valueText("1")
```
Comment (langchain4j) [inline langchain4j-weaviate/src/main/java/dev/langchain4j/store/embedding/weaviate/WeaviateEmbeddingStore.java]: Is it intended that it tries to remove all entries with metadata id 1?

File: langchain4j-weaviate/src/main/java/dev/langchain4j/store/embedding/weaviate/WeaviateMetadataFilterMapper.java [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+import dev.langchain4j.store.embedding.filter.Filter;
+import dev.langchain4j.store.embedding.filter.comparison.*;
+import dev.langchain4j.store.embedding.filter.logical.And;
+import dev.langchain4j.store.embedding.filter.logical.Not;
+import dev.langchain4j.store.embedding.filter.logical.Or;
+import io.weaviate.client.v1.filters.Operator;
+import io.weaviate.client.v1.filters.WhereFilter;
+
+public class WeaviateMetadataFilterMapper {
```
Comment (langchain4j) [inline langchain4j-weaviate/src/main/java/dev/langchain4j/store/embedding/weaviate/WeaviateMetadataFilterMapper.java]:
```suggestion
class WeaviateMetadataFilterMapper {
```

**Comment by langchain4j** [comment]
> Upon debugging, I found that the public String add(Embedding embedding, TextSegment textSegment) method in WeaviateEmbeddingStore doesn't store the provided metadata in Weaviate.

Are you sure?
I have just run `EmbeddingStoreIT.should_add_embedding_with_segment_with_metadata` for `LocalWeaviateEmbeddingStoreIT`, it calls `WeaviateEmbeddingStore.add(Embedding embedding, TextSegment textSegment)` method and stores metadata

File: langchain4j-weaviate/src/main/java/dev/langchain4j/store/embedding/weaviate/WeaviateEmbeddingStore.java (hunk 2 above)
Comment (kiratkumar47) [reply langchain4j-weaviate/src/main/java/dev/langchain4j/store/embedding/weaviate/WeaviateEmbeddingStore.java]: Sorry I hardcoded it for testing, I have updated the code.

**Comment by kiratkumar47** [comment]
I am debugging the `remove_all_by_filter` test in both `LocalWeaviateEmbeddingStoreRemoveIT` & `PgVectorEmbeddingStoreRemoveIT`

In LocalWeaviateEmbeddingStoreRemoveIT I am getting empty metadata 
image

While in PgVectorEmbeddingStoreRemoveIT I am getting 1 metadata with Id as key.
image

**Comment by langchain4j** [comment]
@kiratkumar47 this is a bug then, can you pinpoint the problem please?

**Comment by kiratkumar47** [comment]
It wasn't storing `id` in metadata because `id` was not included in `metadataKeys`. Now, when I add `id` to `metadataKeys`, it stores `id` in metadata. 

I am unable to pass the `remove_all_by_filter` test. I have tried many different approaches, but nothing seems to work. The only method I found is to first fetch the object ID based on the filter, and then delete the record using that object ID. I would really appreciate it if you could guide me on how to do this effectively.

File: langchain4j-weaviate/src/main/java/dev/langchain4j/store/embedding/weaviate/WeaviateEmbeddingStore.java [hunk 4]
```diff
@@ -146,6 +149,45 @@ public List<String> addAll(List<Embedding> embeddings, List<TextSegment> embedde
...
+                        .operator(Operator.ContainsAny)
+                        .valueText(ids.toArray(new String[0]))
+                        .build())
+                .run();
+    }
+
+    @Override
+    public void removeAll(Filter filter) {
+        ensureNotNull(filter, "filter");
+        WhereFilter whereFilter = WeaviateMetadataFilterMapper.map(filter);
```
Comment (langchain4j) [inline langchain4j-weaviate/src/main/java/dev/langchain4j/store/embedding/weaviate/WeaviateEmbeddingStore.java]: I suspect that the problem is that the filter path is "id" but probably should be "_metadata.id" or something like that. Metadata fields are stored not under the root, but under the `_metadata` field

Comment (kiratkumar47) [reply langchain4j-weaviate/src/main/java/dev/langchain4j/store/embedding/weaviate/WeaviateEmbeddingStore.java]: https://github.com/kiratkumar47/langchain4j/blob/3c1a30fccd9c6c69f0d8268ecc5da9b0bf3ec7bc/langchain4j-weaviate/src/main/java/dev/langchain4j/store/embedding/weaviate/WeaviateMetadataFilterMapper.java#L41-L47
I have tried the following parameters in the path, but none of these seem to be working.
1. path("_metadata" + isEqualTo.key())
2. path("/_metadata/" + isEqualTo.key())
3. path("_metadata").path(isEqualTo.key())

image
This is the definition of path in Documentation

Comment (langchain4j) [reply langchain4j-weaviate/src/main/java/dev/langchain4j/store/embedding/weaviate/WeaviateEmbeddingStore.java]: Did you try `path(METADATA, isEqualTo.key())` ?

Comment (langchain4j) [reply langchain4j-weaviate/src/main/java/dev/langchain4j/store/embedding/weaviate/WeaviateEmbeddingStore.java]: METADATA = "_metadata" (just in case)

Comment (kiratkumar47) [reply langchain4j-weaviate/src/main/java/dev/langchain4j/store/embedding/weaviate/WeaviateEmbeddingStore.java]: No working.

Comment (langchain4j) [reply langchain4j-weaviate/src/main/java/dev/langchain4j/store/embedding/weaviate/WeaviateEmbeddingStore.java]: I have created a free weaviate cloud cluster and here is what I see:

image

It seems that metadata field **_(and it's subfields)_** should be filterable for the test to work.

Comment (kiratkumar47) [reply langchain4j-weaviate/src/main/java/dev/langchain4j/store/embedding/weaviate/WeaviateEmbeddingStore.java]: How can I make it Filterable? I couldn't find anything in the documentation.

Comment (langchain4j) [reply langchain4j-weaviate/src/main/java/dev/langchain4j/store/embedding/weaviate/WeaviateEmbeddingStore.java]: Maybe you can ask here: https://forum.weaviate.io/

Comment (kiratkumar47) [reply langchain4j-weaviate/src/main/java/dev/langchain4j/store/embedding/weaviate/WeaviateEmbeddingStore.java]: Yes, I have already posted this issue on both the Weaviate form and StackOverflow.

**Comment by langchain4j** [comment]
Hi @kiratkumar47, what do you think about de-scoping `removeAll(Filter)` from this PR and address it in a separate one (to unblock it)?

**Comment by kiratkumar47** [comment]
> Hi @kiratkumar47, what do you think about de-scoping `removeAll(Filter)` from this PR and address it in a separate one (to unblock it)?

Hi @langchain4j , Sounds great. Should I revert the code for `removeAll` and `WeaviateMetadataFilterMapper` class and comment the test?

**Comment by langchain4j** [comment]
Yes please and I will merge it

**Comment by kiratkumar47** [comment]
I have reverted the `removeAll(Filter filter)` method

**Comment by langchain4j** [comment]
Thanks! Please update docu page as well: https://docs.langchain4j.dev/integrations/embedding-stores/

**Comment by langchain4j** [comment]
No worries, I will update docu

**Comment by langchain4j** [comment]
@kiratkumar47 `removeAll()` seems to not work, could you please check?

You can remove empty disabled `should_remove_all` in `LocalWeaviateEmbeddingStoreRemovalIT` to reproduce.

**Comment by kiratkumar47** [comment]
> @kiratkumar47 `removeAll()` seems to not work, could you please check?
> 
> You can remove empty disabled `should_remove_all` in `LocalWeaviateEmbeddingStoreRemovalIT` to reproduce.

Hi @langchain4j,
Sorry for the late reply, I tried and this method in not working, so I read the documentation and couldn't find anything on removing all objects, the only way I could find is to first fetch all objects and then delete them by ID.

**Comment by langchain4j** [comment]
@kiratkumar47 maybe someon on weaviate forum could help?

**Pull request merged**